File: 20180616-gilded-rose-python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            update(item)


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


def update_before(self):
    if self.brie:
        self.quality += 1


def update(self):
    self.conjured = self.name == "Conjured"
    self.brie = self.name == "Aged Brie"
    self.timeless = self.name == "Sulfuras, Hand of Ragnaros"
    self.backstage = self.name == "Backstage passes to a TAFKAL80ETC concert"
    self.normal = not self.brie and not self.backstage and not self.timeless and not self.conjured

    update_before(self)

    if not self.timeless:
        self.sell_in -= 1

    update_backstage(self)

    if self.sell_in < 0:
        update_after_expiration(self)

    update_normal(self)
    update_conjured(self)
    self.quality = min(50, max(0, self.quality))


def update_normal(self):
    if self.normal:
        by = 2 if self.sell_in < 0 else 1
        self.quality -= by


def update_conjured(self):
    if self.conjured:
        by = 4 if self.sell_in < 0 else 2
        self.quality -= by


def update_after_expiration(self):
    if self.brie:
        self.quality += 1


def update_backstage(self):
    if self.backstage:
        by = 3 if self.sell_in < 5 else 2 if self.sell_in < 10 else 1
        self.quality += by
        if self.sell_in < 0:
            self.quality = 0

File: 20200620-Rose-Scala/src/main/scala/com/gildedrose/GildedRose.scala
package com.gildedrose

class GildedRose(val items: Array[Item]) {
  import GildedRose._

  def updateQuality() {
    items.foreach {
      case i@Item("Sulfuras, Hand of Ragnaros")                => i
      case i@Item("Aged Brie")                                 => i.ageItem().incQuality().incQuality(_.expired)
      case i@Item("Backstage passes to a TAFKAL80ETC concert") => i.ageItem().incQuality().incQuality(_.sellIn < 10).incQuality(_.sellIn < 5).zeroQuality(_.expired)
      case i                                                   => i.ageItem().decQuality().decQuality(_.expired)
    }
  }
}

object GildedRose {
  implicit class ItemOps(val item: Item) extends AnyVal {
    def decQuality(guard: Item => Boolean = _ => true):Item = {
      if (item.quality > 0 && guard(item)) item.quality -= 1
      item
    }
    def incQuality(guard: Item => Boolean = _ => true):Item = {
      if (item.quality < 50 && guard(item)) item.quality += 1
      item
    }
    def expired: Boolean = item.sellIn < 0

    def zeroQuality(guard: Item => Boolean = _ => true):Item = {
      if (guard(item)) item.quality = 0
      item
    }

    def ageItem(): Item = {
      item.sellIn -= 1
      item
    }
  }
}